contracts/inventory/AchievementFacet.sol

```diff
@@ -0,0 +1,307 @@
+
+        emit ItemMarkedAsEquippableInSlot(
+            newSlot,
+            LibInventory.ERC1155_ITEM_TYPE,
+            address(this),
+            poolId,
+            1
+        );
+
+        emit SlotCreated(msg.sender, newSlot);
+        emit NewSlotURI(newSlot);
+        emit NewSlotPersistence(newSlot, true);
+
+        return newSlot;
+    }
+
+    function _mintToInventory(uint256 subjectTokenId, uint256 poolId) internal {
+        LibInventory.InventoryStorage storage istore = LibInventory
+            .inventoryStorage();
+
+        // We could also revert the transaction and say the subject already has this achievement. Or we
+        // could do nothing and assume the achievement is correctly appplied already.
+        if (
+            istore
+            .EquippedItems[istore.ContractERC721Address][subjectTokenId][poolId]
+                .ItemType != 0
+        ) {
+            _unequip(subjectTokenId, poolId, true, 0);
+        }
+
+        ITerminus terminus = ITerminus(address(this));
+        terminus.mint(address(this), poolId, 1, "");
+
+        istore.EquippedItems[istore.ContractERC721Address][subjectTokenId][
+            poolId
+        ] = EquippedItem({
+            ItemType: LibInventory.ERC1155_ITEM_TYPE,
+            ItemAddress: address(this),
+            ItemTokenId: poolId,
+            Amount: 1
+        });
+
+        emit ItemEquipped(
+            subjectTokenId,
+            poolId,
+            LibInventory.ERC1155_ITEM_TYPE,
+            address(this),
+            poolId,
+            1,
+            msg.sender
+        );
+    }
+
+    function adminBatchMintToInventory(
+        uint256[] memory subjectTokenIds,
+        uint256[] memory poolIds
+    ) public {
+        _enforceIsAdmin();
+
+        require(
+            subjectTokenIds.length == poolIds.length,
+            "AchievementFacet.adminBatchMintToInventory: subjectTokenIds and poolIds length mismatch"
+        );
+
+        uint256 i = 0;
+        for (i = 0; i < subjectTokenIds.length; i++) {
+            _mintToInventory(subjectTokenIds[i], poolIds[i]);
+        }
+    }
+
+    function mintHash(
+        uint256 subjectTokenID,
+        uint256 poolID
+    ) public view returns (bytes32) {
+        bytes32 structHash = keccak256(
+            abi.encode(
+                keccak256(
+                    "MintToInventoryMessage(uint256 subjectTokenID,uint256 poolID)"
+                ),
+                subjectTokenID,
+                poolID
+            )
+        );
+        return LibSignatures._hashTypedDataV4(structHash);
+    }
+
+    function mintToInventory(
+        uint256 subjectTokenID,
+        uint256 poolID,
+        address signer,
+        bytes memory signature
+    ) public {
+        LibInventory.InventoryStorage storage istore = LibInventory
+            .inventoryStorage();
+        IERC721 nft = IERC721(istore.ContractERC721Address);
+
+        require(
+            msg.sender == nft.ownerOf(subjectTokenID),
```

Comment:
Do we want this? Don't we want it so that we can mint achievements to an NFT? We won't be owners of that NFT.

Reply:
I see, you are leaning on `adminBatchMintToInventory` for that. I guess I'm ok with this. Wouldn't mind it being permissionless, though, as long as the `msg.sender` has a valid signature.